File: 2023/03/cpp/03.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long          ll;
typedef vector<int>        vi;
typedef vector<bool>       vb;
typedef pair<int, int>     ii;

const int INF = 0x3f3f3f3f;
const ll  INFLL = 0x3f3f3f3f3f3f3f3fll;

#define each(x, s)  for(auto& x : s)
#define loop(x)     for(int i = 0;i < x;i++)
#define vloop(v, x) for(int v = 0;v < x;v++)
#define iter(a)     a.begin(), a.end()
#define riter(a)    a.rbegin(), a.rend()
#define endl        "\n"
#define avg(l, r)   (l + r) / 2

int task1(vector<string> input) {
  int ans = 0;
  int h = input.size();
  int w = input[0].length();
  int val = 0;
  bool part = false;
  loop(h) vloop(j, w) {
    char c = input[i][j];
    if (!isdigit(c)) {
      if (part) ans += val;
      val = 0;
      part = false;
    }
    else {
      val *= 10;
      val += c - '0';
      for (int di = -1;di <= 1;di++) for (int dj = -1;dj <= 1;dj++) {
        if (!di and !dj) continue;
        int ni = i + di;
        int nj = j + dj;
        if (ni < 0 or ni >= h or nj < 0 or nj >= w) continue;
        char c2 = input[ni][nj];
        if (!isdigit(c2) and c2 != '.' and val) {
          part = true;
        }
      }
    }
  }
  return ans;
}

int task2(vector<string> input) {
  int ans = 0;
  int h = input.size();
  int w = input[0].length();
  vector<vi> ids(h, vi(w, -1));
  vi vals;
  int next = 0;
  int val = 0;
  loop(h) vloop(j, w) {
    char c = input[i][j];
    if (!isdigit(c)) {
      for (int k = j - 1;k >= 0 and isdigit(input[i][k]);k--) {
        ids[i][k] = next;
      }
      next++;
      vals.push_back(val);
      val = 0;
    }
    else {
      val *= 10;
      val += c - '0';
    }
  }
  loop(h) vloop(j, w) {
    if (input[i][j] != '*') continue;
    set<int> adj;
    for (int di = -1;di <= 1;di++) for (int dj = -1;dj <= 1;dj++) {
      if (!di and !dj) continue;
      int ni = i + di;
      int nj = j + dj;
      if (ni < 0 or ni >= h or nj < 0 or nj >= w) continue;
      int x = ids[ni][nj];
      if (x >= 0) adj.insert(x);
    }
    if (adj.size() == 2) {
      int p = 1;
      each(x, adj) p *= vals[x];
      ans += p;
    }
  }
  return ans;
}

int main() {
  ios_base::sync_with_stdio(false);cin.tie(NULL);cout.tie(NULL);

  vector<string> input;
  string s;
  while (cin >> s) input.push_back(s + '.');
  cout << "Task 1: " << task1(input) << endl;
  cout << "Task 2: " << task2(input) << endl;
  return 0;
}

File: 2023/03/rust/03.rs
use std::io;

fn isdigit(c: char) -> bool {
    c >= '0' && c <= '9'
}

fn task1(input: &Vec<String>) -> i32 {
    let mut ans = 0;
    let h = input.len() as i32;
    let w = input[0].len() as i32;
    let mut val = 0;
    let mut part = false;
    for i in 0..h {
        for j in 0..w {
            let c = input[i as usize].chars().nth(j as usize).unwrap();
            if !isdigit(c) {
                if part {
                    ans += val;
                }
                val = 0;
                part = false;
            }
            else {
                val *= 10;
                val += c as i32 - '0' as i32;
                for di in -1..=1 {
                    for dj in -1..=1 {
                        if di == 0 && dj == 0 {
                            continue;
                        }
                        let ni = i + di;
                        let nj = j + dj;
                        if ni < 0 || ni >= h || nj < 0 || nj >= w {
                            continue;
                        }
                        let c2 = input[ni as usize].chars().nth(nj as usize).unwrap();
                        if !isdigit(c2) && c2 != '.' && val > 0 {
                            part = true;
                        }
                    }
                }
            }
        }
    }
    ans
}

fn task2(input: &Vec<String>) -> i32 {
    let mut ans = 0;
    let h = input.len() as i32;
    let w = input[0].len() as i32;
    let mut ids = vec![vec![-1; w as usize]; h as usize];
    let mut vals = Vec::new();
    let mut next = 0;
    let mut val = 0;
    for i in 0..h {
        for j in 0..w {
            let c = input[i as usize].chars().nth(j as usize).unwrap();
            if !isdigit(c) {
                for k in (0..j).rev() {
                    if !isdigit(input[i as usize].chars().nth(k as usize).unwrap()) {
                        break;
                    }
                    ids[i as usize][k as usize] = next;
                }
                next += 1;
                vals.push(val);
                val = 0;
            }
            else {
                val *= 10;
                val += c as i32 - '0' as i32;
            }
        }
    }
    for i in 0..h {
        for j in 0..w {
            if input[i as usize].chars().nth(j as usize).unwrap() != '*' {
                continue;
            }
            let mut adj = std::collections::HashSet::<i32>::new();
            for di in -1..=1 {
                for dj in -1..=1 {
                    if di == 0 && dj == 0 {
                        continue;
                    }
                    let ni = i + di;
                    let nj = j + dj;
                    if ni < 0 || ni >= h || nj < 0 || nj >= w {
                        continue;
                    }
                    let x = ids[ni as usize][nj as usize];
                    if x >= 0 {
                        adj.insert(x);
                    }
                }
            }
            if adj.len() == 2 {
                let mut p = 1;
                for x in adj {
                    p *= vals[x as usize];
                }
                ans += p;
            }
        }
    }
    ans
}

fn main() {
    let mut input = Vec::<String>::new();
    let mut line = String::new();
    while io::stdin().read_line(&mut line).unwrap() > 0 {
        line.pop();
        line += ".";
        input.push(line.clone());
        line.clear();
    }
    println!("Task 1: {}", task1(&input));
    println!("Task 2: {}", task2(&input));
}
